### ISSUE TITLE
Logical NOT operator returns wrong boolean literal types with index access

### ISSUE THREAD

**Issue by falsandtru** [Microsoft/TypeScript#28184]
**TypeScript Version:**  3.2.0-dev.201xxxxx

**Search Terms:**

**Code**

**Expected behavior:**

```ts
const a = !Array<object>()[0]; // boolean
const b = !document.children[0]; // boolean
switch (true) {
  case !document.children[0]: // no error
}
switch (false) {
  case !!document.children[0]: // no error
}
```

**Actual behavior:**

```ts
const a = !Array<object>()[0]; // false
const b = !document.children[0]; // false
switch (true) {
  case !document.children[0]: // error
}
switch (false) {
  case !!document.children[0]: // error
}
```

**Playground Link:** 

**Related Issues:**

**Comment by NN---** [comment]
Without strictNullChecks your code is working as you expect.

With strictNullChecks for 'const' variables compiler infers 'false' literal type from the expressions.
If you don't like it, either specify explicitly boolean type or use let.
https://basarat.gitbooks.io/typescript/docs/types/literal-types.html

**Comment by weswigham** [comment]
Root cause is #13778; so I'll mark this as a duplicate of it. Please discuss there~

Duplicates #13778

**Comment by falsandtru** [comment]
@weswigham I have no objection to #13788 and I think we can separate/fix this problem without considering #13778. Can you treat this issue as a special case of logical NOT operator?

**Comment by weswigham** [comment]
No? The cause is simply that the type of, eg `Array<object>()[0]` is always truthy when the index type (ie, `object`) is always truthy, because the indexers do not include `undefined` in their possible result types. Since it's always truthy, the negation of an always-truthy thing is simply `false`, hence the issue at hand.
